### PR TITLE
[WIP] Structured events (NLog 4.5 - alpha5)

### PR DESCRIPTION
**Pull request by 304NotModified**
see https://github.com/NLog/NLog/issues/2077

supersedes https://github.com/NLog/NLog/pull/1951

### CODE REVIEW

**Comment by snakefoot** [comment]
Why not like this:

```
TypeCode objTypeCode = Convert.GetTypeCode(value);
if (objTypeCode == TypeCode.Object && inJson5Mode)
{
   try
   {
      return new ObjectSerializer(this).SerializeObjectProperties(value, objectsInPath, depth, format);
   }
   catch
   {
      return null;
   }
}
else
{
   bool encodeStringValue;
   string escapeXmlString = JsonStringEncode(value, objTypeCode, escapeUnicode, out encodeStringValue);
   if (escapeXmlString != null && encodeStringValue)
      return string.Concat("\"", escapeXmlString, "\"");
   else
      return escapeXmlString;
}
```

**Comment by 304NotModified** [comment]
yes, I will first do  #2128  and then this. 

Currently this feature is too big for me to work on (as you can see....). So first the smaller steps.

File: src/NLog/LogEventInfo.cs [hunk 1]
```diff
@@ -107,19 +105,18 @@ public LogEventInfo(LogLevel level, string loggerName, IFormatProvider formatPro
...
             this.Level = level;
             this.LoggerName = loggerName;
             this.Message = message;
             this.Parameters = parameters;
             this.FormatProvider = formatProvider;
             this.Exception = exception;
-         
+
             if (NeedToPreformatMessage(parameters))
             {
```
Comment (NLog-readonly) [inline src/NLog/LogEventInfo.cs]: MAJOR Either remove or fill this block of code. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS108)

File: src/NLog/Targets/DefaultJsonSerializer.cs [hunk 2]
```diff
@@ -732,5 +733,18 @@ private static PropertyInfo[] GetPropertyInfosNoCache(Type type)
...
 
+        #region Implementation of ISerializer
+
+        /// <summary>Serialize an object</summary>
+        /// <param name="sb">Add serialized value to this builder</param>
+        /// <param name="value">Value to be serialized</param>
+        /// <param name="formatProvider">format</param>
+        public void SerializeObject(StringBuilder sb, object value, IFormatProvider formatProvider)
+        {
+            //todo inject options somewhere else
```
Comment (NLog-readonly) [inline src/NLog/Targets/DefaultJsonSerializer.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/NLog/Logger.cs [hunk 3]
```diff
@@ -403,10 +407,57 @@ private LogEventInfo PrepareLogEventInfo(LogEventInfo logEvent)
...
+
             return logEvent;
 
         }
 
+        /// <summary>
+        /// Parse and render message template.
+        /// </summary>
+        /// <param name="logEvent"></param>
+        private static void HandleMessageTemplate(LogEventInfo logEvent)
```
Comment (NLog-readonly) [inline src/NLog/Logger.cs]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3776)

**Comment by NLog-readonly** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by 304NotModified** [comment]
@snakefoot I change the parse and render of the message template to Logger (after your suggestion, as far as I remember well ;)), but this seems to be a bit problematic - when I change the Message of LogeventInfo, I cannot (re)render the template. 

I think inside logEventInfo was maybe the best place, as there is also an preformMessage there.

Please your thoughts 

Broke also other stuff, just ignore it ;)

**Comment by 304NotModified** [comment]
superseded by https://github.com/NLog/NLog/pull/2208

**Pull request closed without merging**